// livereload-soa/lib/service.ts
'use strict';

export interface CarrierFactory {
    pipe(service: Service): Carrier;
}
export interface Carrier {
    send(message);
}
export interface Service {
    onmessage(message);
    ondisconnect();
}

// A helper base class to make writing services easier.
export class ServiceImpl implements Service {
    private carrier: Carrier;
    private pendingCallbacks = {};
    private nextCallbackId = 1;
    public name: string;
    public options;

    constructor(carrier: CarrierFactory, options) {
        this.options = options;
        this.name = options.name;

        this.initialize();

        this.carrier = carrier.pipe(this);
        this.send({command: 'hello', services: [this.name]});
    }

    initialize() {
    }

    send(message) {
        this.carrier.send(message);
    }

    sendWithReply(message, callback) {
        var callbackId = this.nextCallbackId++;
        this.pendingCallbacks[callbackId] = callback;

        message.reply = { service: this.name, callback: callbackId };
        this.send(message);
    }

    onmessage(message) {
        var callbackId = message.callback;
        if (typeof(callbackId) === 'number') {
            var callback = this.pendingCallbacks[callbackId];
            if (callback) {
                delete this.pendingCallbacks[callbackId];
                callback(message);
            }
        } else if (message.command) {
            var handler = this['on_' + message.command];
            if (!handler)
                throw new Error("Unknown command " + message.command)

            handler.call(this, message, (error, result) => {
                if (message.reply) {
                    var reply = message.reply;
                    if (error)
                        reply.error = error;
                    else
                        reply.result = result;
                    this.send(reply);
                } else if (error) {
                    throw error;
                }
            });
        }
    }

    ondisconnect() {
    }
}
