// livereload-soa/Gruntfile.js
module.exports = function(grunt) {
  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),

    typescript: {
      lib: {
        src: ['src/*.ts'],
        dest: 'lib/',
        options: {
          base_path: 'src',
          module: 'commonjs',
          target: 'es5',
          declaration: true
        }
      },
      example: {
        src: ['example/*.ts'],
        dest: 'example/',
        options: {
          base_path: 'example',
          module: 'commonjs',
          target: 'es5'
        }
      }
    },
    typescript_export: {
      lib: {
        src: ['lib/*.d.ts'],
        dest: 'index.d.ts'
      }
    },
    coffee: {
      lib: {
        expand: true,
        cwd: 'src',
        src: '*.coffee',
        dest: 'lib/',
        ext: '.js'
      }
    },
    simplemocha: {
      options: {
        reporter: 'spec'
      },

      all: { src: ['test/*.js'] }
    }
  });

  grunt.loadNpmTasks('grunt-typescript');
  grunt.loadNpmTasks('grunt-typescript-export');
  grunt.loadNpmTasks('grunt-contrib-coffee');
  grunt.loadNpmTasks('grunt-simple-mocha');

  grunt.registerTask('default', ['typescript:lib', 'typescript_export', 'typescript:example', 'coffee', 'simplemocha']);

  grunt.registerMultiTask('tsexport', 'Concat TypeScript typings for export', function () {
  });
};
