// livereload-soa/Gruntfile.js
module.exports = function(grunt) {
  grunt.initConfig({
    typescript: {
      lib: {
        src: ['src/*.ts'],
        dest: 'lib/',
        options: {
          base_path: 'src',
          module: 'commonjs',
          target: 'es5',
          declaration: true
        }
      },
      examples: {
        src: ['examples/*.ts'],
        dest: 'examples/',
        options: {
          base_path: 'examples',
          module: 'commonjs',
          target: 'es5'
        }
      }
    },
    typescript_export: {
      lib: {
        src: ['lib/*.d.ts'],
        dest: 'index.d.ts'
      }
    },
    simplemocha: {
      options: {
        reporter: 'spec'
      },

      all: { src: ['test/*.js'] }
    }
  });

  grunt.loadNpmTasks('grunt-typescript');
  grunt.loadNpmTasks('grunt-typescript-export');
  grunt.loadNpmTasks('grunt-simple-mocha');

  grunt.registerTask('default', ['typescript:lib', 'typescript_export', 'typescript:examples', 'simplemocha']);

  grunt.registerMultiTask('tsexport', 'Concat TypeScript typings for export', function () {
  });
};
