// livereload-core/lib/rules/compilation-action.js
// Generated by IcedCoffeeScript 1.8.0-d
(function() {
  var Action, CompilationAction,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  Action = require('./action');

  module.exports = CompilationAction = (function(_super) {
    __extends(CompilationAction, _super);

    CompilationAction.prototype.type = 'compile-file';

    function CompilationAction(compiler) {
      this.compiler = compiler;
      CompilationAction.__super__.constructor.call(this, "compile-" + this.compiler.id, "Compile " + this.compiler.name);
    }

    CompilationAction.prototype.createDefaultRules = function() {
      return [
        {
          src: '**/*.' + this.compiler.extensions[0],
          dst: '**/*.' + this.compiler.destinationExt
        }
      ];
    };

    return CompilationAction;

  })(Action);

}).call(this);
