// livereload-core/lib/rules/rule.js
// Generated by IcedCoffeeScript 1.8.0-d
(function() {
  var FileToFileRule, Rule;

  Rule = (function() {
    function Rule() {}

    Rule.prototype.initialize = function(options) {
      return this.setInfo(options);
    };

    Rule.prototype.memento = function() {
      var memento;
      memento = this.getInfo();
      memento.action = this.action.id;
      return memento;
    };

    return Rule;

  })();

  FileToFileRule = (function() {
    function FileToFileRule() {}

    FileToFileRule.prototype.schema = {
      $extends: 'Rule',
      action: {
        type: Object
      },
      sourceSpec: {
        type: String
      },
      destSpec: {
        type: String
      }
    };

    FileToFileRule.prototype.setInfo = function(info) {
      this.sourceSpec = info.src;
      return this.destSpec = info.dst;
    };

    FileToFileRule.prototype.getInfo = function() {
      return {
        src: this.sourceSpec,
        dst: this.destSpec
      };
    };

    return FileToFileRule;

  })();

  exports.Rule = Rule;

  exports.FileToFileRule = FileToFileRule;

}).call(this);
