// livereload-soa/src/api.ts
// a message is a dictionary; some keys are treated specially by various parts of the system
export interface Message {
    // service name to route the message to; only 'hello' message can omit this
    service?: string;

    // a command name; not part of the spec, but a typical way to specify the command
    // (as currently implemented, callback messages omit this and use 'callback' key instead)
    command?: string;
    // not part of the spec; callback ID to invoke, used only in reply messages
    callback?: number;

    // a template for the reply message (currently contains 'service' and 'callback' keys);
    // the spec is that if this key is specified, the reply must be sent after this message is processed
    reply?: Message;

    // one of these is currently appended to the reply message
    error?: Error
    result?: any

    // ... (any further data can be used as message arguments)
}

export interface Carrier {
    send(message: Message): void;
}

export interface MasterCarrier extends Carrier {
    // connects a service that will receive messages, returns a carrier for the service to send the messages via
    pipe(service: Service): Carrier;
}

export interface Service {
    onmessage(message: Message): void;
    ondisconnect(): void;
}
