// livereload-soa/src/muxer.ts
import api = require('./api');

class MuxerService {
    private serviceCarriers = [];

    constructor(public carrier) {
        this.carrier = carrier;
    }

    onmessage(message) {
        var service = this._findService(message);
        if (service)
            service.onmessage(message);
        else
            throw new Error("Cannot route message");
    }

    ondisconnect() {
        this.serviceCarriers.forEach((carrier) => {
            carrier.service.ondisconnect();
        });
    }

    send(message) {
        var service = this._findService(message);
        if (service)
            service.onmessage(message);
        else
            this.carrier.send(message);
    }

    pipe(service) {
        var carrier = new MuxerCarrier(this, service);
        this.serviceCarriers.push(carrier);
        return carrier;
    }

    updateProvidedServices() {
        var providedServices = [];
        this.serviceCarriers.forEach((carrier) => {
            providedServices.push.apply(providedServices, carrier.providedServices);
        });
        this.send({ command: 'hello', services: providedServices });
    }

    _findService(message) {
        var service = null;
        this.serviceCarriers.forEach((carrier) => {
            if (carrier.providedServices.indexOf(message.service) >= 0) {
                service = carrier.service;
                // break (if we could)
            }
        });
        return service;
    }
}

class MuxerCarrier {
    public providedServices = [];

    constructor(public muxer, public service) {
    }

    send(message) {
        if (message.command === 'hello') {
            this.providedServices = message.services;
            this.muxer.updateProvidedServices();
        } else {
            this.muxer.send(message);
        }
    }
}

export function createMuxer(carrier): api.Service {
    return new MuxerService(carrier);
}
